### ISSUE TITLE
DAOImpl.findOptionalById shouldn't be final

### ISSUE THREAD

**Issue by bigtrouble** [jOOQ/jOOQ#12811]
Hi
I tried to upgrade to JOOQ 3.16.1 today. 
But I found a problem，I'm not sure if it's a bug.

I know the method `findOptionalById` was added in DAOImpl.
```java
    @Override
    public final Optional<P> findOptionalById(T id) {
        return Optional.ofNullable(findById(id));
    }
```

When I generate the code through jooq-codegen and run it something like this:
```
CglibAopProxy:262-Unable to proxy interface-implementing method 
[public final java.util.Optional org.jooq.impl.DAOImpl.findOptionalById(java.lang.Object)] 
because it is marked as final: Consider using interface-based JDK proxies instead!
```

My JOOQ-CodeGen config like this:
```xml
    <generate>
      <relations>true</relations>
      <daos>true</daos>
      <springAnnotations>true</springAnnotations>
      ....  omit other ...
    </generate>
```

It seems that Spring(cglib) has some conflict with dao's findOptionalById，
and removing the `final` flag solves the problem.

### Versions:

- jOOQ: 3.16.1 
- Java: 11
- Springboot: 2.2.6.RELEASE
- Database (include vendor): MySQL 5.7.20
- OS: Linux
- JDBC Driver (include name if inofficial driver): -

**Comment by lukaseder** [comment]
Thanks a lot for your report. `DAOImpl` is one of those rare exceptions where we don't make methods final, so this is clearly a bug. Will fix right away.

**Comment by lukaseder** [comment]
`DAOImpl::ctx` is also `final`, but shouldn't be.

**Comment by lukaseder** [comment]
I guess the `DAOImpl::ctx` method didn't cause any issues pre jOOQ 3.16, because it isn't declared in the `DAO` interface?

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.2 (https://github.com/jOOQ/jOOQ/issues/12812), and 3.15.6 (https://github.com/jOOQ/jOOQ/issues/12813, `ctx()` method only).

Thanks again for your report.
